Lib/operator.py

```diff
@@ -408,6 +410,56 @@ def ixor(a, b):
     return a
 
 
+# Asynchronous Iterator Operations ********************************************#
+
+
+_NOT_PROVIDED = object()  # sentinel object to detect when a kwarg was not given
+
+
+def aiter(obj, sentinel=_NOT_PROVIDED):
+    """aiter(async_iterable) -> async_iterator
+    aiter(async_callable, sentinel) -> async_iterator
+
+    Like the iter() builtin but for async iterables and callables.
+    """
+    if sentinel is _NOT_PROVIDED:
+        if not isinstance(obj, AsyncIterable):
+            raise TypeError(f'aiter expected an AsyncIterable, got {type(obj)}')
+        if isinstance(obj, AsyncIterator):
+            return obj
+        return (i async for i in obj)
```

Comment:
```suggestion
        ait = obj.__aiter__()
        if not hasattr(ait, '__anext__'):
            raise TypeError(f"aiter() returned non-AsyncIterator of type {type(ait)!r}")
        return ait
```
I think it should be this way for a few reasons:

1. It eliminates the two separate non-raising cases and return statements/types
2. It eliminates the overhead of the extra async generator expression and is more easily implemented in C.
4. For any AsyncIterator `a`, we "should" have `a.__aiter__() is a`, so the behavior shouldn't change much.
5. For the cases where that isn't true, in analogy with the builtin `iter()`, one might expect `aiter(x)` to always call `x.__aiter__()`:

```python
from collections.abc import Iterator, AsyncIterator

class A:
    def __next__(self):
        return 1
    def __iter__(self):
        return iter([])

# "a" is already an iterator, but a.__iter__ is called anyway.
a = A()
assert isinstance(a, Iterator)
assert iter(a) is not a
assert type(iter(a)) is not type(a)
assert type(iter(a)) is type(iter([]))

############################

async def another_aiter():
    yield 1

class B:
    def __anext__(self):
        return 1
    def __aiter__(self):
        return another_aiter()

b = B()
assert isinstance(b, AsyncIterator)
# These three assertions fail, but pass with the change.
assert aiter(b) is not b
assert type(aiter(b)) is not type(b)
assert type(aiter(b)) is type(aiter(another_aiter()))
```

See the implementation of `iter()` [here](https://github.com/python/cpython/blob/master/Objects/abstract.c#L2616), which uses `PyObject_GetIter()` from [here](https://github.com/python/cpython/blob/master/Python/bltinmodule.c#L1511).
I wonder if a `PyObject_GetAsyncIter` needs to be added to the C-API so that `aiter()` and [the GET_AITER opcode](https://github.com/python/cpython/blob/master/Python/ceval.c#L2196) can use the same code.

Reply:
Shouldn't we call `type(obj).__aiter__(obj)` (and the analogue for `__anext__`)? That's how dunder methods are generally called by the interpreter.

Reply:
Thanks for reviewing! Updated this patch based on the good suggestions.

(Note, I used `isinstance(ait, AsyncIterator)` rather than the `hasattr(ait, '__anext__')` that was suggested, figuring that's the preferred spelling. Also, after rebasing on latest master, I had to move the `from collections.abc ...` imports from module scope to function scope as a workaround to avoid a circular dependency between `operator` and `collections`, which wasn't an issue when I first submitted this PR. Sounds like this PR will now serve as a reference implementation for a reimplementation in C (submitted in a separate PR) anyway, where I guess such a workaround won't be necessary?)